File: solutions/0022_Generate_Parentheses/backtracking.java
class Solution {
    public List<String> generateParenthesis(int n) {
        return recur("", n, 0, new ArrayList<String>());
    }

    private List<String> recur(String s, int n, int m, List<String> list) {
        if (n < 0 || m < 0) return list;
        if (n == 0 && m == 0) {
            list.add(s);
        }
        recur(s + "(", n - 1, m + 1, list);
        recur(s + ")", n, m - 1, list);
        return list;
    }
}

File: solutions/0022_Generate_Parentheses/backtracking.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def recur(p, n, m):
            if n >= 0 and m >= 0:
                # print(p, n, m)
                if n is 0 and m is 0:
                    yield p
                for q in recur(p + '(', n - 1, m + 1): yield q
                for q in recur(p + ')', n, m - 1): yield q
        return list(recur('', n, 0))
